src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,378 @@
+                    )
+
+            if not is_text:
+                return builder.strategy.map(lambda u: u.encode('utf-8'))
+            return builder.strategy
+
+        elif code == sre.ANY:
+            # Regex '.' (any char)
+            if is_text:
+                if context.flags & re.DOTALL:
+                    return hs.characters()
+                return hs.characters(blacklist_characters=u'\n')
+            else:
+                if context.flags & re.DOTALL:
+                    return binary_char
+                return binary_char.filter(lambda c: c != b'\n')
+
+        elif code == sre.AT:
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
+            # An empty string (or newline) will match the token itself, but
+            # we don't and can't check the position (eg '%' at the end)
+            if value == sre.AT_END:
+                return hs.one_of(hs.just(EMPTY),
+                                 hs.just(u'\n' if is_text else b'\n'))
+            return hs.just(EMPTY)
+
+        elif code == sre.SUBPATTERN:
+            # Various groups: '(...)', '(:...)' or '(?P<name>...)'
+            old_flags = context.flags
+            if HAS_SUBPATTERN_FLAGS:  # pragma: no cover
+                # This feature is available only in specific Python versions
+                context.flags = (context.flags | value[1]) & ~value[2]
+
+            strat = _strategy(value[-1], context, is_text)
+
+            context.flags = old_flags
+
+            if value[0]:
+                context.groups[value[0]] = strat
+                strat = hs.shared(strat, key=value[0])
+
+            return strat
+
+        elif code == sre.GROUPREF:
+            # Regex '\\1' or '(?P=name)' (group reference)
+            return hs.shared(context.groups[value], key=value)
+
+        elif code == sre.ASSERT:
+            # Regex '(?=...)' or '(?<=...)' (positive lookahead/lookbehind)
+            return recurse(value[1])
+
+        elif code == sre.ASSERT_NOT:
+            # Regex '(?!...)' or '(?<!...)' (negative lookahead/lookbehind)
+            return hs.just(EMPTY)
+
+        elif code == sre.BRANCH:
+            # Regex 'a|b|c' (branch)
+            return hs.one_of([recurse(branch) for branch in value[1]])
+
+        elif code in [sre.MIN_REPEAT, sre.MAX_REPEAT]:
+            # Regexes 'a?', 'a*', 'a+' and their non-greedy variants
+            # (repeaters)
+            at_least, at_most, subregex = value
+            if at_most == sre.MAXREPEAT:
+                at_most = None
+            if at_least == 0 and at_most == 1:
+                return hs.just(EMPTY) | recurse(subregex)
+            return hs.lists(recurse(subregex),
+                            min_size=at_least,
+                            max_size=at_most).map(EMPTY.join)
+
+        elif code == sre.GROUPREF_EXISTS:
+            # Regex '(?(id/name)yes-pattern|no-pattern)'
+            # (if group exists choice)
+            return hs.one_of(
```

Comment:
I don't think this is correct is it? We have access to whether the group exists at this point so we should use it.